secrets/provider/kubernetes/provider.go

```diff
@@ -55,25 +75,58 @@ func (p k8sProvider) CleanupModel(*provider.ModelBackendConfig) error {
 	return nil
 }
 
-func (p k8sProvider) getBroker(cfg *provider.ModelBackendConfig) (Broker, cloudspec.CloudSpec, error) {
-	cloudSpec, err := p.configToCloudSpec(&cfg.BackendConfig)
+func (p k8sProvider) getBroker(cfg *provider.ModelBackendConfig) (*kubernetesClient, error) {
+	validCfg, err := newConfig(cfg.Config)
 	if err != nil {
-		return nil, cloudspec.CloudSpec{}, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	envCfg, err := config.New(config.UseDefaults, map[string]interface{}{
-		config.NameKey: cfg.ModelName,
-		config.UUIDKey: cfg.ModelUUID,
-		config.TypeKey: state.ModelTypeCAAS,
-	})
+	restCfg, err := configToK8sRestConfig(validCfg)
 	if err != nil {
-		return nil, cloudspec.CloudSpec{}, errors.Trace(err)
+		return nil, errors.Annotatef(err, "invalid k8s config")
+	}
+	k8sClient, err := NewK8sClient(restCfg)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	broker := &kubernetesClient{
+		namespace:         validCfg.namespace(),
+		isControllerModel: cfg.ModelName == bootstrap.ControllerModelName,
+		client:            k8sClient,
+	}
+	return broker, errors.Trace(err)
+}
+
+func configToK8sRestConfig(cfg *backendConfig) (*rest.Config, error) {
+	if cfg.useInClusterAddress() {
+		rc, err := InClusterConfig()
+		if err != nil && !errors.Is(err, rest.ErrNotInCluster) {
+			return nil, errors.Trace(err)
+		}
+		if rc != nil {
+			logger.Tracef("using in-cluster config")
+			return rc, nil
+		}
```

Comment:
Do we need to raise an error if it's not in-cluster but the user requested to use in-cluster?

Reply:
Ah, the method should have been called `preferInClusterAddress()`. We don't have the necessary modelling to allow this to be done properly. The only time this is even used is when creating the back end to correspond to a cloud added to the controller (it uses the value of "is-controller-cloud"=true/false). It was added for microk8s. When using k8s as an external backend, code is never run as the user always needs to specify a correct externally reachable IP address.